pyiceberg/table/name_mapping.py

```diff
@@ -0,0 +1,203 @@
+    def fields(self, struct: List[MappedField], field_results: List[T]) -> T:
+        """Visit a List[MappedField]."""
+
+    @abstractmethod
+    def field(self, field: MappedField, field_result: T) -> T:
+        """Visit a MappedField."""
+
+
+class _IndexById(NameMappingVisitor[Dict[int, MappedField]]):
+    result: Dict[int, MappedField]
+
+    def __init__(self) -> None:
+        self.result = {}
+
+    def mapping(self, nm: NameMapping, field_results: Dict[int, MappedField]) -> Dict[int, MappedField]:
+        return field_results
+
+    def fields(self, struct: List[MappedField], field_results: List[Dict[int, MappedField]]) -> Dict[int, MappedField]:
+        return self.result
+
+    def field(self, field: MappedField, field_result: Dict[int, MappedField]) -> Dict[int, MappedField]:
+        if field.field_id in self.result:
+            raise ValueError(f"Invalid mapping: ID {field.field_id} is not unique")
+
+        self.result[field.field_id] = field
+
+        return self.result
+
+
+class _IndexByName(NameMappingVisitor[Dict[str, MappedField]]):
+    def mapping(self, nm: NameMapping, field_results: Dict[str, MappedField]) -> Dict[str, MappedField]:
+        return field_results
+
+    def fields(self, struct: List[MappedField], field_results: List[Dict[str, MappedField]]) -> Dict[str, MappedField]:
+        return dict(ChainMap(*field_results))
+
+    def field(self, field: MappedField, field_result: Dict[str, MappedField]) -> Dict[str, MappedField]:
+        result: Dict[str, MappedField] = {
+            f"{field_name}.{key}": result_field for key, result_field in field_result.items() for field_name in field.names
+        }
+
+        for name in field.names:
+            result[name] = field
+
+        return result
+
+
+@singledispatch
+def visit_name_mapping(obj: Union[NameMapping, List[MappedField], MappedField], visitor: NameMappingVisitor[T]) -> T:
+    """Traverse the name mapping in post-order traversal."""
+    raise NotImplementedError(f"Cannot visit non-type: {obj}")
+
+
+@visit_name_mapping.register(NameMapping)
+def _(obj: NameMapping, visitor: NameMappingVisitor[T]) -> T:
+    return visitor.mapping(obj, visit_name_mapping(obj.root, visitor))
+
+
+@visit_name_mapping.register(list)
+def _(fields: List[MappedField], visitor: NameMappingVisitor[T]) -> T:
+    results = [visitor.field(field, visit_name_mapping(field.fields, visitor)) for field in fields]
+    return visitor.fields(fields, results)
+
+
+def load_mapping_from_json(mapping: str) -> NameMapping:
+    return NameMapping.model_validate_json(mapping)
+
+
+class _CreateMapping(SchemaVisitor[List[MappedField]]):
+    def schema(self, schema: Schema, struct_result: List[MappedField]) -> List[MappedField]:
+        return struct_result
+
+    def struct(self, struct: StructType, field_results: List[List[MappedField]]) -> List[MappedField]:
+        return [
+            MappedField(field_id=field.field_id, names={field.name}, fields=result)
```

Comment:
```suggestion
            MappedField(field_id=field.field_id, names=[field.name], fields=result)
```
This applies to other similar instances as well

Reply:
Good one, thanks!